addons/web/static/tests/fields/relational_fields_tests.js

```diff
@@ -3702,6 +3707,80 @@ QUnit.module('relational_fields', {
 
         form.destroy();
     });
+    QUnit.test('Check onchange with two consecutive nested many2one', async function (assert) {
+        assert.expect(5);
+        this.data.turtle.onchanges = {
+            turtle_trululu: function (record) {
+                if (record.turtle_trululu === 1) {
+                    record.o = [37];
+                    record.p = [17, 19];
+                }
+                else {
+                    record.o = [41];
+                    record.p = [19]; 
+                }
+            },
+        };
+        let onchange_count = 0;
+        var form = await createView({
+            View: FormView,
+            model: 'turtle',
+            data: this.data,
+            arch: 
+                '<form string="Turtles">' +
+                    '<field string="Product" name="turtle_trululu"/>' +
+                    '<field readonly="1" string="Related field" name="o">' +
+                        '<tree>' +
+                            '<field name="partner"/>' +
+                        '</tree>' +
+                    '</field>' +
+                    '<field readonly="1" string="Second related field" name="p">' +
+                        '<tree>' +
+                            '<field name="partner_ids"/>' +
+                        '</tree>' +
+                    '</field>' +
+                '</form>',
+            res_id: 1,
+            mockRPC: function (route, args) {
+                if (args.method === 'onchange') {
+                    if (!onchange_count) {
+                        assert.deepEqual(args.args[1], {
+                            'turtle_trululu': 1,
+                            id: 1,
+                            o: [],
+                            p: [],
+                        });
+                    } else if (onchange_count === 1) {
+                        assert.deepEqual(args.args[1], {
+                            'turtle_trululu': 2,
+                            id: 1,
+                            o: [
+                                [4, 37, false],
+                            ],
+                            p: [
+                                [4, 17, false],
+                                [4, 19, false],
+                            ],
+                        });
+                    }
+                    onchange_count++;
+                    assert.step(args.method + ' ' + args.model);
+                }
+                return this._super(route, args);
+            }
+        });
+
+        await testUtils.form.clickEdit(form);
+        await testUtils.fields.many2one.clickOpenDropdown("turtle_trululu");
+        await testUtils.fields.many2one.searchAndClickItem('turtle_trululu', {search: 'first record'});
+        await testUtils.fields.many2one.clickOpenDropdown("turtle_trululu");
+        await testUtils.fields.many2one.searchAndClickItem('turtle_trululu', {search: 'second record'});
```

Comment:
Could you quickly try to add widget="many2many_tags" on the 2 inner one2manys, only do the first onchange and assert that both many2many_tags avec fetched the display_names (so only check they are displayed correctly). If I understand correctly, one of them should not display the display_names without you fix. No need to do the second onchange then